giraf/binaryGiraf/binReader.go

```diff
@@ -0,0 +1,301 @@
+	pathLen := binary.LittleEndian.Uint16(br.currData.Next(2)) // pathLen (uint16)
+	var node uint32
+	for i = 0; i < pathLen; i++ { // path ([]uint32)
+		node = binary.LittleEndian.Uint32(br.currData.Next(4))
+		answer.Path.Nodes = append(answer.Path.Nodes, node)
+	}
+
+	// cigar ([]cigar.ByteCigar)
+	numCigarOps := binary.LittleEndian.Uint16(br.currData.Next(2)) // numCigarOps (uint16)
+	for i = 0; i < numCigarOps; i++ {
+		answer.Cigar = append(answer.Cigar, cigar.ByteCigar{
+			RunLen: binary.LittleEndian.Uint16(br.currData.Next(2)), // byteCigar.RunLen (uint16)
+			Op:     br.currData.Next(1)[0],                          // byteCigar.Op (byte)
+		})
+	}
+
+	// fancySeq (dnaThreeBit.ThreeBit)
+	var fancyBases dnaThreeBit.ThreeBit
+	fancyBases.Len = int(binary.LittleEndian.Uint32(br.currData.Next(4))) // fancySeq.Len (uint32)
+
+	// figure out how many uint64 are needed to store fancySeqLen worth of bases
+	//TODO: Craig, do you think this is the best way to determine this?
+	numInts := fancyBases.Len / 21
+	if fancyBases.Len%21 != 0 {
+		numInts++
+	}
+
+	for j := 0; j < numInts; j++ {
+		fancyBases.Seq = append(fancyBases.Seq, binary.LittleEndian.Uint64(br.currData.Next(8))) // fancySeq.Seq (uint64)
+	}
+	addFullSeq(&answer, &fancyBases, g)
+
+	// alnScore (int64)
+	answer.AlnScore = int(binary.LittleEndian.Uint64(br.currData.Next(8)))
+
+	// mapQ (uint8)
+	answer.MapQ = br.currData.Next(1)[0]
+
+	// qual ([]cigar.ByteCigar)
+	var op uint8
+	numQualOps := binary.LittleEndian.Uint16(br.currData.Next(2)) // numQualOps (uint16)
+	for i = 0; i < numQualOps; i++ {
+		runLen = binary.LittleEndian.Uint16(br.currData.Next(2)) // byteCigar.RunLen (uint16)
+		op = br.currData.Next(1)[0]                              // byteCigar.Op (byte)
+		for j = 0; j < runLen; j++ {
+			answer.Qual = append(answer.Qual, op)
+		}
+	}
+
+	// QStart, QEnd (int)
+	answer.QStart, answer.QEnd = determineQStartQEnd(&answer)
+
+	// notes ([]BinNote)
+	appendNotes(&answer, br) // notes ([]bytes)
+
+	if giraf.IsForwardRead(&answer) {
+		answer.PosStrand = true
+	}
+
+	return answer, err
+}
+
+// addFullSeq parses the cigar and the fancySeq fields to retrieve the full length read sequence
+func addFullSeq(answer *giraf.Giraf, fancySeq *dnaThreeBit.ThreeBit, graph *simpleGraph.SimpleGraph) {
+	var fancyBases []dna.Base
+	if fancySeq.Len != 0 {
+		fancyBases = dnaThreeBit.ToDnaBases(fancySeq)
+	}
+	refIdx := answer.Path.TStart
+	var currNodeId int
+	var currNode *simpleGraph.Node = graph.Nodes[answer.Path.Nodes[0]]
+	var i uint16
+
+	for _, cigar := range answer.Cigar {
+		switch cigar.Op {
+		case '=':
+			for i = 0; i < cigar.RunLen; i++ {
+				if refIdx > len(currNode.Seq)-1 {
+					refIdx = 0
+					currNodeId++
+					currNode = graph.Nodes[answer.Path.Nodes[currNodeId]]
+				}
+
+				answer.Seq = append(answer.Seq, currNode.Seq[refIdx])
+				refIdx++
+			}
+		case 'X':
+			answer.Seq = append(answer.Seq, fancyBases[:cigar.RunLen]...)
+			fancyBases = fancyBases[:cigar.RunLen]
```

Comment:
I have somehow confused myself.  I was picturing this line having "cigar.RunLen:" instead of ":cigar.RunLen".  I was thinking that you were processing the first cigar.RunLen bases from fancyBases (on the previous line) and then removing those bases on this line.  Maybe a few comments would help me?